File: 46-permutations/permutations.cpp
class Solution {
public:
    vector<vector<int>>  result;

    void backtrack(vector<int>& nums,vector<int> curr){
        if(curr.size()==nums.size()){
            result.push_back(curr);
            return;
        }
        for(int num:nums){
            if(find(curr.begin(),curr.end(),num)==curr.end()){
                curr.push_back(num);
                backtrack(nums,curr);
                curr.pop_back();
            }
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        backtrack(nums,{});
        return result;
    }
};

File: 46-permutations/permutations.java
class Solution {
    List<List<Integer>> result=new ArrayList<>();

    public void backtrack(int[] nums,List<Integer> curr){
        if(curr.size()==nums.length){
            result.add(new ArrayList<>(curr));
            return;
        }
        for(int num:nums){
            if(!curr.contains(num)){
                curr.add(num);
                backtrack(nums,curr);
                curr.remove(curr.size()-1);
            }
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        backtrack(nums,new ArrayList<>());
        return result;
    }
}